source/agora/test/Base.d

```diff
@@ -907,7 +907,103 @@ public class TestAPIManager
+        string file = __FILE__, size_t line = __LINE__)
+    {
+        generateBlocks(target_height, enrolled_height, clients.take(GenesisValidators),
+            file, line);
+    }
+
+    /// Ditto
+    public void generateBlocks (Clients)(Height target_height,
+        Height enrolled_height = Height(0),
+        Clients clients = this.clients,
+        string file = __FILE__, size_t line = __LINE__)
+    {
+        assert(target_height >= enrolled_height,
+            format!"[%s:%s] Target height %s is not at least enroll height %s"
+                (file, line, target_height, enrolled_height));
+        iota(target_height - clients[0].getAllBlocks()[$ -1].header.height)
+            .each!(_ => addBlock(enrolled_height, clients, file, line));
+    }
+
+     /*******************************************************************************
+
+        Add a block
+
+        Params:
+            target_height = the target height after adding blocks.
+            enrolled_height = the height with the active enrollments
+            client_index = the client to use for putting the tx
+
+        Returns:
+            The number of blocks added to reach the target height
+
+    *******************************************************************************/
+    void addBlock (Clients)(Height enrolled_height = Height(0),
+        Clients clients = this.clients,
+        string file = __FILE__, size_t line = __LINE__)
+    {
+        auto all_blocks = clients[0].getAllBlocks();
+        Height target_height = Height(all_blocks[$ - 1].header.height + 1);
+        assert(all_blocks.length > 0,
+            format!"[%s:%s] No blocks in first client!"(file, line));
+        auto spendables = all_blocks[$ - 1].spendable().array;
+        auto tx_count = max(1, this.test_conf.txs_to_nominate);
+        assert(spendables.length >= tx_count,
+            format!"[%s:%s] Less than %s spendables in block:\n%s"
+                (file, line, tx_count, prettify(all_blocks[$ - 1])));
+        spendables.takeExactly(tx_count)
+            .map!(txb => txb.sign())
+            .each!(tx => clients[0].putTransaction(tx));
+        ushort distance = cast(ushort) (target_height - enrolled_height -1);
+        assert(distance < GenesisValidatorCycle && distance >= 0,
+            format!"[%s:%s] Expected distance between 0 and %s not %s"
+                (file, line, GenesisValidatorCycle - 1, distance));
+        expectBlock(clients, target_height, all_blocks[enrolled_height].header);
+    }
+
+    /*******************************************************************************
+
+        Enroll validator
+
+        Params:
+            client_idx = the index of the client to enroll
+
+    *******************************************************************************/
+
+    public void enroll (size_t client_idx,
+        string file = __FILE__, size_t line = __LINE__)
+    {
+        enroll(this.clients, client_idx, file, line);
```

Comment:
`this.enroll`

Reply:
Although technically you could just make the other overload static I suppose, since you're already giving it the range of clients.

Reply:
Refactored these functions which were taking clients as a template parameter to just take an array with the indices of the clients which are the participating validators. Defaulting to the Genesis enrolled validators.